src/backend/distributed/commands/utility_hook.c

```diff
@@ -1533,6 +1564,97 @@ NodeDDLTaskList(TargetWorkerSet targets, List *commands)
 }
 
 
+static void
+SplitCoordinatorJob(DDLJob *ddlJob, DDLJob **coordinatorDDLJob,
+					DDLJob **nonCoordinatorNodesDDLJob)
+{
+	*coordinatorDDLJob = palloc(sizeof(DDLJob));
+	*nonCoordinatorNodesDDLJob = palloc(sizeof(DDLJob));
+
+	**coordinatorDDLJob = *ddlJob;
+	**nonCoordinatorNodesDDLJob = *ddlJob;
+
+	(*coordinatorDDLJob)->taskList = NIL;
+	(*nonCoordinatorNodesDDLJob)->taskList = NIL;
+
+	Task *task = NULL;
+	foreach_ptr(task, ddlJob->taskList)
+	{
+		Task *coordinatorTask = NULL;
+		Task *nonCoordinatorNodesTask = NULL;
+		SplitCoordinatorTask(task, &coordinatorTask, &nonCoordinatorNodesTask);
+
+		if (coordinatorTask != NULL)
+		{
+			(*coordinatorDDLJob)->taskList =
+				lappend((*coordinatorDDLJob)->taskList, coordinatorTask);
+		}
+
+		if (nonCoordinatorNodesTask != NULL)
+		{
+			(*nonCoordinatorNodesDDLJob)->taskList =
+				lappend((*nonCoordinatorNodesDDLJob)->taskList, nonCoordinatorNodesTask);
+		}
+	}
+
+	if ((*coordinatorDDLJob)->taskList == NIL)
+	{
+		*coordinatorDDLJob = NULL;
+	}
+
+	if ((*nonCoordinatorNodesDDLJob)->taskList == NIL)
+	{
+		*nonCoordinatorNodesDDLJob = NULL;
+	}
+
+	if (*coordinatorDDLJob && *nonCoordinatorNodesDDLJob)
+	{
+		/* defer the metadata sync to "other nodes" phase */
```

Comment:
Afaict #7016 would not be enough to solve the problem. My thinking is that if we execute `standardProcessUtility` first on the current node, that will actually take locks on distributed shell table on the current node first (which could be a worker). And instead we want those locks to be taken on the coordinator first to avoid deadlocks. Sorting the workers where to send DDL would prevent deadlocks when actually changing the shards, but it will not prevent deadlocks that occur when changing the distributed shell table.

Reply:
> To avoid the deadlocks, we could be more aggressive such that we acquire a global advisory lock on the first worker node per OID of the object changed. 

This would help avoiding from the deadlocks for the cases we acquire this adv lock. However, say if distributed_ddl (that's allowed from any node) can get into a distributed deadlock with citus_operation() (that's only allowed from coordinator), then we need to not forget about acquiring this adv lock in those operations too.